codegen/src/main/ops/org/nd4j/codegen/ops/SDBaseOps.kt

```diff
@@ -0,0 +1,1699 @@
+        Arg(BOOL, "exclusive") { description = "If true: exclude the first value" }
+        Arg(BOOL, "reverse") { description = "If true: reverse the direction of the accumulation" }
+        Arg(INT, "axis") { count = AtLeast(1); description = "Scalar axis argument for dimension to perform cumululative sum operations along" }
+        Output(NUMERIC, "output"){ description = "Output variable" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Cumulative product operation.
+                For input: [ a, b, c], output is:
+                exclusive=false, reverse=false: [a, a*b, a*b*c]
+                exclusive=true, reverse=false, [0, a, a*b]
+                exclusive=false, reverse=true: [a*b*c, b*c, c]
+                exclusive=true, reverse=true: [b*c, c, 0]
+            """.trimIndent()
+        }
+    }
+
+    Op("cumsum") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.transforms.custom"
+        javaOpClass = "CumSum"
+        Input(NUMERIC, "in") { description = "Input variable" }
+        Arg(BOOL, "exclusive") { description = "If true: exclude the first value" }
+        Arg(BOOL,  "reverse") { description = "If true: reverse the direction of the accumulation" }
+        Arg(INT, "axis") { count = AtLeast(1); description = "Scalar axis argument for dimension to perform cumululative sum operations along" }
+        Output(NUMERIC, "output"){ description = "" }
+
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Cumulative sum operation.
+                For input: [ a, b, c], output is:
+                exclusive=false, reverse=false: [a, a+b, a+b+c]
+                exclusive=true, reverse=false, [0, a, a+b]
+                exclusive=false, reverse=true: [a+b+c, b+c, c]
+                exclusive=true, reverse=true: [b+c, c, 0]
+            """.trimIndent()
+        }
+    }
+
+    Op("dot") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.reduce3"
+        javaOpClass = "Dot"
+        legacy = true
+        Input(NUMERIC, "x") { description = "first input" }
+        Input(NUMERIC, "y") { description = "second input" }
+        Arg(INT, "dimensions") {count = AtLeast(1); description = "Dimensions to reduce over. If dimensions are not specified, full array reduction is performed" }
+        Output(NUMERIC, "output"){ description = "output variable" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Pairwise dot product reduction along dimension
+                output = sum(i=0 ... size(dim)-1) x[i] * y[i]
+            """.trimIndent()
+        }
+    }
+
+    Op("dynamicPartition") {
+        //TODO: double check this one, may have gotten it wrong.
+        javaPackage = "org.nd4j.linalg.api.ops.impl.transforms.custom"
+        javaOpClass = "DynamicPartition"
+        Input(NUMERIC, "x") { description = "Input variable" }
+        Arg(INT, "partitions") { count = AtLeast(1); description = "1D input with values 0 to numPartitions-1" }
```

Comment:
This will also require a new function in NDValidation:

    NDValidation.validateInteger("dynamicPartition", "partitions", partitions);

does not compile.